File: tests/test_api_routes.ts
/**
 * Layer G — API route integration tests.
 *
 * Tests the /api/facilities/[state] public JSON endpoint against a live server
 * (production by default; override with TEST_BASE_URL env var for preview deploys).
 *
 * Every test hits the REAL HTTP endpoint with ZERO mocks. Tests verify:
 *   1. HTTP status codes
 *   2. Response JSON schema completeness (every required field present)
 *   3. Data correctness (known fixtures appear; counts in expected range)
 *   4. CORS and cache headers
 *   5. Error paths (unknown states → 404)
 *
 * Run:
 *   npx tsx tests/test_api_routes.ts
 *   TEST_BASE_URL=https://your-preview.vercel.app npx tsx tests/test_api_routes.ts
 *
 * Exit 0 = all passed. Exit 1 = one or more failed.
 */

import { readFileSync } from "node:fs";
import { resolve } from "node:path";

// ---------------------------------------------------------------------------
// Env loader
// ---------------------------------------------------------------------------

function loadEnv(): void {
  const repoRoot = resolve(__dirname, "..");
  for (const name of [".env.local", ".env"]) {
    try {
      const content = readFileSync(resolve(repoRoot, name), "utf8");
      for (const line of content.split("\n")) {
        const trimmed = line.trim();
        if (!trimmed || trimmed.startsWith("#")) continue;
        const eqIdx = trimmed.indexOf("=");
        if (eqIdx === -1) continue;
        const key = trimmed.slice(0, eqIdx).trim();
        const rawVal = trimmed.slice(eqIdx + 1).trim().replace(/^["']|["']$/g, "");
        if (key && !(key in process.env)) process.env[key] = rawVal;
      }
      break;
    } catch {
      // file not found
    }
  }
}

loadEnv();

// ---------------------------------------------------------------------------
// Test runner
// ---------------------------------------------------------------------------

type TestFn = () => Promise<void>;
const tests: Array<{ name: string; fn: TestFn }> = [];
let passed = 0;
let failed = 0;
const failures: string[] = [];

function test(name: string, fn: TestFn): void {
  tests.push({ name, fn });
}

let networkSkipped = false;

async function runAll(): Promise<void> {
  console.log("=".repeat(60));
  console.log("StarlynnCare — Layer G: API Route Integration Tests");
  console.log(`Target: ${BASE_URL}`);
  console.log("=".repeat(60));

  for (const t of tests) {
    if (networkSkipped) {
      console.log(`  SKIP  ${t.name}  (network unavailable)`);
      continue;
    }
    try {
      await t.fn();
      console.log(`  PASS  ${t.name}`);
      passed++;
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      // If this is a network error, skip all remaining tests
      if (msg.includes("Network unavailable")) {
        networkSkipped = true;
        console.log(`  SKIP  ${t.name}  (network unavailable — run from dev machine with internet access)`);
        console.log(`        ${msg.split("\n")[0]}`);
        continue;
      }
      console.log(`  FAIL  ${t.name}`);
      console.log(`        ← ${msg}`);
      failed++;
      failures.push(t.name);
    }
  }

  console.log();
  console.log("-".repeat(60));
  console.log(`Summary: ${passed}/${passed + failed} passed`);
  if (failures.length > 0) {
    console.log(`Failed (${failures.length}):`);
    for (const f of failures) console.log(`  • ${f}`);
  }
  console.log("-".repeat(60));
}

function assert(condition: boolean, message: string): void {
  if (!condition) throw new Error(message);
}

// ---------------------------------------------------------------------------
// Config
// ---------------------------------------------------------------------------

const BASE_URL = (process.env.TEST_BASE_URL ?? "https://starlynn.care").replace(/\/$/, "");

// Covered state slugs (from states.ts COVERED_STATES)
const COVERED_SLUGS = ["california", "texas", "oregon", "washington", "minnesota", "utah", "illinois"] as const;

// StateFacility fields that must be present on every facility in the API response
// (from the ApiPayload type in /api/facilities/[state]/route.ts)
const REQUIRED_FACILITY_FIELDS = [
  "id",
  "name",
  "url",
  "state_code",
  "care_category",
  "serves_memory_care",
  "total_deficiency_count",
] as const;

// Required top-level ApiPayload fields
const REQUIRED_PAYLOAD_FIELDS = [
  "schema",
  "state_code",
  "state_name",
  "generated_at",
  "count",
  "methodology_url",
  "facilities",
] as const;

// ---------------------------------------------------------------------------
// Helper
// ---------------------------------------------------------------------------

async function getJson(path: string): Promise<{ status: number; headers: Headers; body: unknown }> {
  const url = `${BASE_URL}${path}`;
  let response: Response;
  try {
    response = await fetch(url, {
      headers: { Accept: "application/json" },
      // 30-second timeout
      signal: AbortSignal.timeout(30_000),
    });
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    if (
      msg.includes("fetch failed") ||
      msg.includes("ECONNREFUSED") ||
      msg.includes("ENOTFOUND") ||
      msg.includes("network") ||
      msg.includes("timeout") ||
      msg.includes("AbortError")
    ) {
      throw new Error(
        `Network unavailable: ${msg}\n` +
        `URL: ${url}\n` +
        `Set TEST_BASE_URL to a reachable server, or run from a machine with internet access.`,
      );
    }
    throw err;
  }
  let body: unknown;
  try {
    body = await response.json();
  } catch {
    body = null;
  }
  return { status: response.status, headers: response.headers, body };
}

// ---------------------------------------------------------------------------
// Tests: unknown / invalid states return 404
// ---------------------------------------------------------------------------

test("unknown state returns 404", async () => {
  const { status } = await getJson("/api/facilities/atlantis");
  assert(status === 404, `Expected 404 for unknown state, got ${status}`);
});

test("pennsylvania returns 404 (not in COVERED_STATES)", async () => {
  const { status } = await getJson("/api/facilities/pennsylvania");
  assert(
    status === 404,
    `Expected 404 for 'pennsylvania' (not in COVERED_STATES), got ${status}. ` +
    "PA has DB data but is not yet exposed via the API.",
  );
});

test("empty state slug returns 404", async () => {
  const { status } = await getJson("/api/facilities/");
  // Could be 404 or 405 depending on routing
  assert(status >= 400, `Expected ≥400 for empty state slug, got ${status}`);
});

// ---------------------------------------------------------------------------
// Tests: each covered state returns valid payload
// ---------------------------------------------------------------------------

for (const stateSlug of COVERED_SLUGS) {
  test(`${stateSlug}: returns HTTP 200`, async () => {
    const { status } = await getJson(`/api/facilities/${stateSlug}`);
    assert(status === 200, `Expected 200, got ${status}`);
  });

  test(`${stateSlug}: response has required top-level fields`, async () => {
    const { body } = await getJson(`/api/facilities/${stateSlug}`);
    const payload = body as Record<string, unknown>;
    assert(payload !== null && typeof payload === "object", "Response is not a JSON object");

    for (const field of REQUIRED_PAYLOAD_FIELDS) {
      assert(field in payload, `Missing top-level field: '${field}'`);
    }
  });

  test(`${stateSlug}: schema field is schema.org Dataset`, async () => {
    const { body } = await getJson(`/api/facilities/${stateSlug}`);
    const payload = body as Record<string, unknown>;
    assert(
      payload.schema === "https://schema.org/Dataset",
      `Expected schema='https://schema.org/Dataset', got '${payload.schema}'`,
    );
  });

  test(`${stateSlug}: count matches facilities array length`, async () => {
    const { body } = await getJson(`/api/facilities/${stateSlug}`);
    const payload = body as Record<string, unknown>;
    const facilities = payload.facilities as unknown[];
    assert(Array.isArray(facilities), "facilities must be an array");
    assert(
      payload.count === facilities.length,
      `count=${payload.count} but facilities.length=${facilities.length}`,
    );
  });

  test(`${stateSlug}: has at least one facility`, async () => {
    const { body } = await getJson(`/api/facilities/${stateSlug}`);
    const payload = body as Record<string, unknown>;
    const facilities = payload.facilities as unknown[];
    assert(
      facilities.length > 0,
      `${stateSlug}: 0 facilities returned. Ingest or recompute has not run.`,
    );
  });

  test(`${stateSlug}: every facility has required fields`, async () => {
    const { body } = await getJson(`/api/facilities/${stateSlug}`);
    const payload = body as Record<string, unknown>;
    const facilities = (payload.facilities as Array<Record<string, unknown>>).slice(0, 10);

    for (const facility of facilities) {
      for (const field of REQUIRED_FACILITY_FIELDS) {
        assert(
          field in facility,
          `Facility '${facility.name}' missing field '${field}'`,
        );
      }
      // serves_memory_care must be boolean
      assert(
        typeof facility.serves_memory_care === "boolean",
        `serves_memory_care is not boolean for '${facility.name}'`,
      );
      // URL must be a valid https URL
      assert(
        typeof facility.url === "string" && facility.url.startsWith("https://"),
        `url is not https:// for '${facility.name}': ${facility.url}`,
      );
      // total_deficiency_count must be a number
      assert(
        typeof facility.total_deficiency_count === "number",
        `total_deficiency_count is not a number for '${facility.name}'`,
      );
    }
  });

  test(`${stateSlug}: facility URL path matches state slug`, async () => {
    const { body } = await getJson(`/api/facilities/${stateSlug}`);
    const payload = body as Record<string, unknown>;
    const facilities = (payload.facilities as Array<Record<string, unknown>>).slice(0, 5);

    for (const facility of facilities) {
      const url = facility.url as string;
      assert(
        url.includes(`/${stateSlug}/`),
        `Facility URL '${url}' doesn't contain state slug '/${stateSlug}/'`,
      );
    }
  });

  test(`${stateSlug}: generated_at is a valid ISO datetime`, async () => {
    const { body } = await getJson(`/api/facilities/${stateSlug}`);
    const payload = body as Record<string, unknown>;
    const generatedAt = payload.generated_at as string;
    const parsed = new Date(generatedAt);
    assert(!isNaN(parsed.getTime()), `generated_at '${generatedAt}' is not a valid datetime`);
    // Must be recent (within 2 hours, accounting for cache)
    const ageMs = Date.now() - parsed.getTime();
    // Allow stale-while-revalidate window (86400 = 1 day)
    assert(ageMs < 90_000_000, `generated_at is over 25 hours old: ${generatedAt}`);
  });
}

// ---------------------------------------------------------------------------
// Tests: California-specific data correctness
// ---------------------------------------------------------------------------

test("california: Opal Care appears in facility list", async () => {
  const { body } = await getJson("/api/facilities/california");
  const payload = body as Record<string, unknown>;
  const facilities = (payload.facilities as Array<Record<string, unknown>>);

  const opalCare = facilities.find(
    (f) =>
      typeof f.license_number === "string" && f.license_number === "200672",
  );
  assert(
    opalCare !== undefined,
    "Opal Care (license 200672) not found in /api/facilities/california. " +
    "This is CA's most-cited facility and a key regression fixture.",
  );
});

test("california: Opal Care has serves_memory_care=true", async () => {
  const { body } = await getJson("/api/facilities/california");
  const payload = body as Record<string, unknown>;
  const facilities = (payload.facilities as Array<Record<string, unknown>>);

  const opalCare = facilities.find(
    (f) => typeof f.license_number === "string" && f.license_number === "200672",
  );
  assert(opalCare !== undefined, "Opal Care not found in API response");
  assert(
    opalCare!.serves_memory_care === true,
    `Opal Care serves_memory_care=${opalCare!.serves_memory_care} in API response`,
  );
});

test("california: Opal Care has total_deficiency_count > 0", async () => {
  const { body } = await getJson("/api/facilities/california");
  const payload = body as Record<string, unknown>;
  const facilities = (payload.facilities as Array<Record<string, unknown>>);

  const opalCare = facilities.find(
    (f) => typeof f.license_number === "string" && f.license_number === "200672",
  );
  assert(opalCare !== undefined, "Opal Care not found in API response");
  assert(
    (opalCare!.total_deficiency_count as number) > 0,
    `Opal Care total_deficiency_count=${opalCare!.total_deficiency_count} — expected > 0`,
  );
});

test("california: count exceeds minimum (>= 1000)", async () => {
  const { body } = await getJson("/api/facilities/california");
  const payload = body as Record<string, unknown>;
  assert(
    (payload.count as number) >= 1000,
    `CA facility count=${payload.count}, expected ≥1000`,
  );
});

// ---------------------------------------------------------------------------
// Tests: CORS and cache headers
// ---------------------------------------------------------------------------

for (const stateSlug of COVERED_SLUGS.slice(0, 2)) {
  test(`${stateSlug}: CORS header allows all origins`, async () => {
    const { headers } = await getJson(`/api/facilities/${stateSlug}`);
    const corsHeader = headers.get("access-control-allow-origin");
    assert(
      corsHeader === "*",
      `Expected Access-Control-Allow-Origin: *, got '${corsHeader}'`,
    );
  });

  test(`${stateSlug}: X-Robots-Tag is noindex (API responses suppressed from web search)`, async () => {
    const { headers } = await getJson(`/api/facilities/${stateSlug}`);
    const robotsHeader = headers.get("x-robots-tag");
    assert(
      robotsHeader !== null && robotsHeader.toLowerCase().includes("noindex"),
      `Expected X-Robots-Tag: noindex, got '${robotsHeader}'`,
    );
  });

  test(`${stateSlug}: Cache-Control header is set`, async () => {
    const { headers } = await getJson(`/api/facilities/${stateSlug}`);
    const cacheControl = headers.get("cache-control");
    assert(
      cacheControl !== null && cacheControl.length > 0,
      `Cache-Control header missing for ${stateSlug}`,
    );
  });
}

// ---------------------------------------------------------------------------
// Tests: methodology_url is a valid URL
// ---------------------------------------------------------------------------

test("methodology_url is a valid https URL", async () => {
  const { body } = await getJson("/api/facilities/california");
  const payload = body as Record<string, unknown>;
  const url = payload.methodology_url as string;
  assert(
    typeof url === "string" && url.startsWith("https://"),
    `methodology_url '${url}' is not a valid https URL`,
  );
});

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

runAll().then(() => {
  if (failed > 0) process.exit(1);
}).catch((err) => {
  console.error("Fatal error:", err);
  process.exit(1);
});

File: tests/integration/api/test_api_routes.py
#!/usr/bin/env python3
"""
Layer 3 (extended) — API route integration tests against the live production URL.

Tests every Next.js API route against the actual deployed production endpoint:

  • GET /api/facilities/{state}  — public JSON-LD dataset (all 10 covered states)
  • GET /api/facilities/invalid  — must return 404 with { error: "..." }
  • GET /api/export/ca-inspections — CA inspection CSV bulk export

No mocks. No mock server. Hits https://www.starlynncare.com directly.

Each facility in each state's response is validated for:
  - Field presence (every field the FacilityRow type declares)
  - Type correctness (booleans, numbers, strings)
  - Value constraints (URL patterns, state codes, positive counts)
  - Cross-consistency (state_code in facility matches route state)

Exit 0 = all checks passed. Exit 1 = one or more checks failed.
"""
from __future__ import annotations

import re
import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[3]))
from tests._lib import (  # noqa: E402
    check,
    run_all_checks,
    make_session,
    PRODUCTION_URL,
    COVERED_STATES,
    MIN_PUBLISHABLE,
    HTTP_TIMEOUT,
)

# ── Expected types + constraints ───────────────────────────────────────────────

# Fields the StateFacility type (in /api/facilities/[state]/route.ts) exposes.
# Each entry: (field_name, expected_python_type, nullable)
FACILITY_FIELDS: list[tuple[str, type | tuple, bool]] = [
    ("id",                    str,         False),
    ("name",                  str,         False),
    ("url",                   str,         False),
    ("street",                str,         True),
    ("city",                  str,         True),
    ("postal_code",           str,         True),
    ("state_code",            str,         False),
    ("latitude",              (int, float), True),
    ("longitude",             (int, float), True),
    ("license_number",        str,         True),
    ("license_type",          str,         True),
    ("beds",                  (int, float), True),
    ("care_category",         str,         False),
    ("serves_memory_care",    bool,        False),
    ("capacity_tier",         str,         True),
    ("regulator_url",         str,         True),
    ("last_inspection_date",  str,         True),
    ("total_deficiency_count", int,        False),
    ("updated_at",            str,         True),
]

# Top-level fields in the API payload
PAYLOAD_FIELDS = [
    "schema", "state_code", "state_name", "generated_at", "count",
    "methodology_url", "facilities",
]

# Known care_category values from the schema
VALID_CARE_CATEGORIES = {
    "assisted_living_memory_care",
    "memory_care_only",
    "nursing_home_memory_care",
    "assisted_living",
    "residential_care",
    "adult_family_home",
    "board_and_care",
}

URL_RE = re.compile(r"^https?://")
ISO_DATE_RE = re.compile(r"^\d{4}-\d{2}-\d{2}")


# ── Helpers ────────────────────────────────────────────────────────────────────

def _validate_facility_object(label: str, fac: dict, expected_state_code: str) -> None:
    """Validate a single facility object from the API response."""
    # Check every declared field is present
    for field_name, field_type, nullable in FACILITY_FIELDS:
        val = fac.get(field_name)
        if not nullable:
            check(
                f"{label}.{field_name}: non-null",
                val is not None,
                f"got None (field is required)",
            )
        if val is not None:
            check(
                f"{label}.{field_name}: correct type",
                isinstance(val, field_type),
                f"expected {field_type}, got {type(val).__name__}={val!r}"[:80],
            )

    # URL must start with https://www.starlynncare.com/
    url = fac.get("url", "")
    check(
        f"{label}.url: starts with production base",
        url.startswith(f"{PRODUCTION_URL}/"),
        f"url={url!r}",
    )

    # state_code must match the route
    sc = fac.get("state_code", "")
    check(
        f"{label}.state_code: matches route ({expected_state_code})",
        sc == expected_state_code,
        f"got {sc!r}",
    )

    # total_deficiency_count must be ≥ 0
    tdc = fac.get("total_deficiency_count")
    if tdc is not None:
        check(
            f"{label}.total_deficiency_count: ≥ 0",
            isinstance(tdc, int) and tdc >= 0,
            f"got {tdc!r}",
        )

    # care_category must be a known value
    cc = fac.get("care_category", "")
    check(
        f"{label}.care_category: recognized value",
        cc in VALID_CARE_CATEGORIES,
        f"got {cc!r} (valid={VALID_CARE_CATEGORIES})",
    )

    # serves_memory_care must be a boolean
    smc = fac.get("serves_memory_care")
    check(
        f"{label}.serves_memory_care: is bool",
        isinstance(smc, bool),
        f"got {type(smc).__name__}={smc!r}",
    )

    # last_inspection_date, if set, must look like an ISO date
    lid = fac.get("last_inspection_date")
    if lid is not None:
        check(
            f"{label}.last_inspection_date: ISO date format",
            bool(ISO_DATE_RE.match(str(lid))),
            f"got {lid!r}",
        )

    # If regulator_url is set, must be an HTTP URL
    reg_url = fac.get("regulator_url")
    if reg_url is not None:
        check(
            f"{label}.regulator_url: is URL",
            bool(URL_RE.match(reg_url)),
            f"got {reg_url!r}",
        )


# ── Tests ──────────────────────────────────────────────────────────────────────

def test_facilities_api_per_state(sess) -> None:
    """GET /api/facilities/{state} for all covered states — full schema validation."""
    print("\n[/api/facilities/{state} — all 10 states]")

    for state_code, state_slug in COVERED_STATES:
        url = f"{PRODUCTION_URL}/api/facilities/{state_slug}"
        print(f"\n  → {state_slug} ({state_code})")

        t0 = time.monotonic()
        r = sess.get(url, timeout=HTTP_TIMEOUT * 2)
        elapsed = time.monotonic() - t0

        label = f"/api/facilities/{state_slug}"
        check(f"{label}: HTTP 200", r.status_code == 200, f"got {r.status_code}")
        if r.status_code != 200:
            continue

        check(
            f"{label}: Content-Type JSON",
            "application/json" in r.headers.get("content-type", ""),
            f"content-type={r.headers.get('content-type')!r}",
        )

        try:
            payload = r.json()
        except Exception as e:
            check(f"{label}: valid JSON", False, str(e))
            continue
        check(f"{label}: valid JSON", True)

        # Top-level payload structure
        for field in PAYLOAD_FIELDS:
            check(f"{label}: payload.{field} present", field in payload, f"missing from response")

        check(
            f"{label}: schema = 'https://schema.org/Dataset'",
            payload.get("schema") == "https://schema.org/Dataset",
            f"got {payload.get('schema')!r}",
        )
        check(
            f"{label}: state_code = '{state_code}'",
            payload.get("state_code") == state_code,
            f"got {payload.get('state_code')!r}",
        )
        check(
            f"{label}: methodology_url is URL",
            bool(URL_RE.match(str(payload.get("methodology_url", "")))),
            f"got {payload.get('methodology_url')!r}",
        )
        check(
            f"{label}: generated_at is ISO timestamp",
            bool(ISO_DATE_RE.match(str(payload.get("generated_at", "")))),
            f"got {payload.get('generated_at')!r}",
        )

        facilities = payload.get("facilities", [])
        count_field = payload.get("count", -1)
        check(
            f"{label}: count field matches facilities array length",
            count_field == len(facilities),
            f"count={count_field} facilities_len={len(facilities)}",
        )

        min_n = MIN_PUBLISHABLE.get(state_code, 1)
        check(
            f"{label}: facility count ≥ {min_n}",
            len(facilities) >= min_n,
            f"got {len(facilities)}",
        )

        # Deep-validate the first 10 facilities and last 5 (catch sort edge cases)
        sample = facilities[:10] + facilities[-5:] if len(facilities) > 10 else facilities
        for i, fac in enumerate(sample):
            is_tail = i >= 10
            idx = len(facilities) - len(facilities[-5:]) + (i - 10) if is_tail else i
            fac_label = f"{label}/fac[{idx}]"
            _validate_facility_object(fac_label, fac, state_code)

        # Aggregate: no duplicate IDs in the response
        all_ids = [f.get("id") for f in facilities]
        unique_ids = set(all_ids)
        check(
            f"{label}: no duplicate facility IDs",
            len(all_ids) == len(unique_ids),
            f"total={len(all_ids)} unique={len(unique_ids)}",
        )

        # Aggregate: all serves_memory_care facilities must have a non-empty name
        mc_facilities = [f for f in facilities if f.get("serves_memory_care")]
        check(
            f"{label}: all memory-care facilities have non-empty names",
            all(bool(f.get("name")) for f in mc_facilities),
            f"mc_count={len(mc_facilities)}",
        )

        print(f"    OK  {len(facilities)} facilities in {elapsed:.1f}s")


def test_invalid_state_returns_404(sess) -> None:
    """GET /api/facilities/{invalid-slug} returns HTTP 404 with error JSON."""
    print("\n[/api/facilities/{state} — invalid state slug → 404]")

    invalid_slugs = ["mars", "canada", "not-a-state", "ca123"]
    for slug in invalid_slugs:
        url = f"{PRODUCTION_URL}/api/facilities/{slug}"
        r = sess.get(url, timeout=HTTP_TIMEOUT)
        check(f"/api/facilities/{slug}: HTTP 404", r.status_code == 404, f"got {r.status_code}")
        if r.status_code == 404:
            try:
                body = r.json()
                check(
                    f"/api/facilities/{slug}: error field present",
                    "error" in body,
                    f"body={body}",
                )
            except Exception:
                pass  # Non-JSON 404 is still a 404


def test_ca_inspections_export(sess) -> None:
    """GET /api/export/ca-inspections returns a CSV response."""
    print("\n[/api/export/ca-inspections — CSV bulk export]")

    url = f"{PRODUCTION_URL}/api/export/ca-inspections"
    r = sess.get(url, timeout=HTTP_TIMEOUT * 3, stream=True)
    check(
        "CA export: HTTP 200",
        r.status_code == 200,
        f"got {r.status_code}",
    )
    if r.status_code != 200:
        return

    ct = r.headers.get("content-type", "").lower()
    check(
        "CA export: Content-Type is text/csv or octet-stream",
        "csv" in ct or "octet-stream" in ct or "text/plain" in ct,
        f"content-type={ct!r}",
    )

    # Read the first 32 KB to validate the CSV header
    chunk = b""
    for piece in r.iter_content(chunk_size=8192):
        chunk += piece
        if len(chunk) >= 32768:
            break
    r.close()

    text = chunk.decode("utf-8-sig", errors="replace")
    import csv, io  # noqa: PLC0415
    reader = csv.reader(io.StringIO(text))
    try:
        header = next(reader)
        check("CA export: has header row", len(header) >= 5, f"got {len(header)} columns")
        # Must have facility-identifying columns
        header_lower = {h.strip().lower() for h in header}
        for expected_col in ("facility", "inspection", "date"):
            check(
                f"CA export: column containing '{expected_col}' present",
                any(expected_col in h for h in header_lower),
                f"header={list(header_lower)[:10]}",
            )
        # Check at least one data row
        data_row = next(reader, None)
        check("CA export: has at least 1 data row", data_row is not None, "empty CSV")
    except StopIteration:
        check("CA export: parseable CSV", False, "no rows found in first 32KB")


def test_cache_control_headers(sess) -> None:
    """API responses include appropriate Cache-Control headers."""
    print("\n[Cache-Control headers on facilities API]")

    url = f"{PRODUCTION_URL}/api/facilities/california"
    r = sess.get(url, timeout=HTTP_TIMEOUT * 2)
    if r.status_code != 200:
        check("Cache-Control check: skipped (endpoint returned error)", True, f"got {r.status_code}")
        return

    cc = r.headers.get("cache-control", "")
    check(
        "CA facilities API: Cache-Control header set",
        bool(cc),
        f"cache-control={cc!r}",
    )
    check(
        "CA facilities API: CORS header present",
        "access-control-allow-origin" in {k.lower() for k in r.headers},
        f"headers={list(r.headers.keys())}",
    )
    check(
        "CA facilities API: X-Robots-Tag=noindex",
        r.headers.get("x-robots-tag", "").lower() == "noindex",
        f"x-robots-tag={r.headers.get('x-robots-tag')!r}",
    )


# ── Main ───────────────────────────────────────────────────────────────────────

def main() -> None:
    print("=" * 60)
    print("StarlynnCare — Layer 3 (Extended): API Route Integration Tests")
    print(f"Target: {PRODUCTION_URL}")
    print("=" * 60)

    sess = make_session()
    sess.headers["Accept"] = "application/json"

    test_facilities_api_per_state(sess)
    test_invalid_state_returns_404(sess)
    test_ca_inspections_export(sess)
    test_cache_control_headers(sess)

    run_all_checks("Layer 3 Extended (API routes)")


if __name__ == "__main__":
    main()
